x/auction/keeper/keeper.go

```diff
@@ -0,0 +1,497 @@
+	}
+
+	// Calculate somm per sale token price
+	// Starting price is amount of usomm required for 1 of starting denom
+	saleTokenPriceInUsomm := lastSaleTokenPrice.UsdPrice.Quo(lastUsommTokenPrice.UsdPrice)
+
+	auctionID := k.GetLastAuctionID(ctx) + 1
+	auction := types.Auction{
+		Id:                         auctionID,
+		StartingTokensForSale:      startingTokensForSale,
+		StartBlock:                 uint64(ctx.BlockHeight()),
+		EndBlock:                   0,
+		InitialPriceDecreaseRate:   initialPriceDecreaseRate,
+		CurrentPriceDecreaseRate:   initialPriceDecreaseRate,
+		PriceDecreaseBlockInterval: priceDecreaseBlockInterval,
+		InitialUnitPriceInUsomm:    saleTokenPriceInUsomm,
+		CurrentUnitPriceInUsomm:    saleTokenPriceInUsomm,
+		RemainingTokensForSale:     startingTokensForSale,
+		FundingModuleAccount:       fundingModuleAccount,
+		ProceedsModuleAccount:      proceedsModuleAccount,
+	}
+
+	if err := auction.ValidateBasic(); err != nil {
+		return err
+	}
+
+	// Validate no ongoing auction for denom
+	var err error = nil
+	k.IterateAuctions(ctx, types.GetActiveAuctionsPrefix(), func(auctionID uint32, auction types.Auction) (stop bool) {
+		if auction.StartingTokensForSale.Denom == startingTokensForSale.Denom {
+			err = sdkerrors.Wrapf(types.ErrCannotStartTwoAuctionsForSameDenomSimultaneously, "Denom: %s", auction.StartingTokensForSale.Denom)
+			return true
+		}
+		return false
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// Transfer the coins
+	if err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, fundingModuleAccount, types.ModuleName, sdk.Coins{startingTokensForSale}); err != nil {
+		return err
+	}
+
+	// Add auction to active auctions
+	k.setActiveAuction(ctx, auction)
+
+	// Update last auction id
+	k.setLastAuctionID(ctx, auctionID)
+
+	// Emit event that auction has started
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeNewAuction,
+			sdk.NewAttribute(types.AttributeKeyAuctionID, fmt.Sprint(k.GetLastAuctionID(ctx))),
+			sdk.NewAttribute(types.AttributeKeyStartBlock, fmt.Sprint(ctx.BlockHeight())),
+			sdk.NewAttribute(types.AttributeKeyInitialDecreaseRate, fmt.Sprintf("%f", initialPriceDecreaseRate)),
+			sdk.NewAttribute(types.AttributeKeyBlockDecreaseInterval, fmt.Sprint(priceDecreaseBlockInterval)),
+			sdk.NewAttribute(types.AttributeKeyStartingDenom, startingTokensForSale.Denom),
+			sdk.NewAttribute(types.AttributeKeyStartingAmount, startingTokensForSale.Amount.String()),
+			sdk.NewAttribute(types.AttributeKeyStartingUsommPrice, saleTokenPriceInUsomm.String()),
+		),
+	)
+
+	return nil
+}
+
+func (k Keeper) tokenPriceTooOld(ctx sdk.Context, tokenPrice *types.TokenPrice) bool {
+	return uint64(ctx.BlockHeight())-tokenPrice.LastUpdatedBlock > k.GetParamSet(ctx).PriceMaxBlockAge
+}
+
+// FinishAuction completes an auction by sending relevant funds to destination addresses and updates state
+func (k Keeper) FinishAuction(ctx sdk.Context, auction *types.Auction) error {
```

Comment:
My understanding, and it may be outdated, is that the keeper.go function should be for "safe" activities against the KV store. Checks for things that may cause errors should be handler before these functions are called. Maybe BeginAuction and FinishAuction should be in a separate file like keeper/auction.go ?

Reply:
I'm definitely not an authoritative source on the matter, but it appears that at least the cosmos bank keeper, for example, does have some functions that can potentially error (e.g. https://github.com/cosmos/cosmos-sdk/blob/main/x/bank/keeper/keeper.go#L364-L366). 

Perhaps @zmanian / @jackzampolin can provide more insight on the best pattern to use in this keeper.go?